File: src/conduit/loader/s3_loader.py
"""S3 destination writer (built-in via PyArrow S3 filesystem)."""

from __future__ import annotations

import logging
from pathlib import Path

import pyarrow as pa
import pyarrow.csv as pcsv
import pyarrow.parquet as pq
from pyarrow import fs as pafs

from conduit.models import DestinationConfig

logger = logging.getLogger(__name__)


def load_s3(table: pa.Table, dest: DestinationConfig, base_dir: Path) -> None:
    """Write an Arrow table to an S3 bucket as CSV or Parquet.

    Uses PyArrow's built-in S3FileSystem — no boto3 dependency required.
    Credentials are resolved from environment variables, IAM roles,
    or explicit config values.
    """
    bucket = dest.config.get("bucket")
    key = dest.config.get("key", dest.config.get("path", f"{dest.name}.parquet"))
    file_format = dest.config.get("format", "parquet")
    region = dest.config.get("region", "")

    if not bucket:
        raise ValueError("S3 destination requires 'bucket' in config")

    # Build S3FileSystem — picks up AWS_* env vars and IAM roles automatically
    s3_kwargs = {}
    if region:
        s3_kwargs["region"] = region
    if dest.config.get("aws_access_key_id"):
        s3_kwargs["access_key"] = dest.config["aws_access_key_id"]
        s3_kwargs["secret_key"] = dest.config.get("aws_secret_access_key", "")
    if dest.config.get("endpoint_override"):
        s3_kwargs["endpoint_override"] = dest.config["endpoint_override"]

    s3 = pafs.S3FileSystem(**s3_kwargs)
    s3_path = f"{bucket}/{key}"

    if file_format == "csv":
        with s3.open_output_stream(s3_path) as out:
            write_options = pcsv.WriteOptions(include_header=True)
            pcsv.write_csv(table, out, write_options=write_options)
    else:
        pq.write_table(
            table,
            s3_path,
            filesystem=s3,
            compression=dest.config.get("compression", "snappy"),
        )

    logger.info(
        "Loaded %d rows to S3 destination '%s' (s3://%s/%s, format=%s)",
        len(table), dest.name, bucket, key, file_format,
    )

File: src/ts/loader/s3_loader.ts
/**
 * S3 destination writer using AWS SDK v3.
 */

import {
  S3Client,
  PutObjectCommand,
} from "@aws-sdk/client-s3";
import { stringify } from "@std/csv";
import { DuckDBInstance } from "@duckdb/node-api";
import type { DataTable, DestinationConfig } from "../models.ts";
import { writeCsvSync } from "../util.ts";

export async function loadS3(
  table: DataTable,
  dest: DestinationConfig,
  _baseDir: string,
): Promise<void> {
  const bucket = dest.config.bucket as string;
  const key = (dest.config.key as string) ??
    (dest.config.path as string) ?? `${dest.name}.parquet`;
  const fileFormat = (dest.config.format as string) ?? "parquet";
  const region = (dest.config.region as string) ?? undefined;

  if (!bucket) {
    throw new Error("S3 destination requires 'bucket' in config");
  }

  const clientConfig: Record<string, unknown> = {};
  if (region) clientConfig.region = region;
  if (dest.config.aws_access_key_id) {
    clientConfig.credentials = {
      accessKeyId: dest.config.aws_access_key_id as string,
      secretAccessKey: (dest.config.aws_secret_access_key as string) ?? "",
    };
  }
  if (dest.config.endpoint_override) {
    clientConfig.endpoint = dest.config.endpoint_override as string;
    clientConfig.forcePathStyle = true;
  }

  const s3 = new S3Client(clientConfig);
  let body: Uint8Array;
  let contentType: string;

  if (fileFormat === "csv") {
    const colNames = table.columns.map((c) => c.name);
    const rows = table.rows.map((row) => {
      const obj: Record<string, string> = {};
      for (const col of colNames) {
        const v = row[col];
        obj[col] = v === null || v === undefined ? "" : String(v);
      }
      return obj;
    });
    const csv = stringify(rows, { columns: colNames });
    body = new TextEncoder().encode(csv);
    contentType = "text/csv";
  } else {
    // Write parquet via DuckDB
    const tmpDir = Deno.makeTempDirSync();
    const tmpPath = `${tmpDir}/_s3_data.parquet`;
    const csvPath = `${tmpDir}/_s3_data.csv`;
    writeCsvSync(csvPath, table);

    const instance = await DuckDBInstance.create();
    const conn = await instance.connect();
    try {
      await conn.run(
        `CREATE TABLE _s3 AS SELECT * FROM read_csv('${csvPath}', auto_detect=true)`,
      );
      const compression = (dest.config.compression as string) ?? "snappy";
      await conn.run(
        `COPY _s3 TO '${tmpPath}' (FORMAT PARQUET, COMPRESSION '${compression}')`,
      );
    } finally {
      conn.closeSync();
    }

    body = Deno.readFileSync(tmpPath);
    contentType = "application/octet-stream";
    try {
      Deno.removeSync(tmpDir, { recursive: true });
    } catch {
      // best effort
    }
  }

  await s3.send(
    new PutObjectCommand({
      Bucket: bucket,
      Key: key,
      Body: body,
      ContentType: contentType,
    }),
  );

  console.log(
    `Loaded ${table.rows.length} rows to S3 destination '${dest.name}' (s3://${bucket}/${key}, format=${fileFormat})`,
  );
}
